modules/telemetry/telemetry/include/hephaestus/telemetry/struclog.h

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <iomanip>
+#include <source_location>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <utility>
+#include <vector>
+
+namespace heph::telemetry {
+
+enum class Level : std::uint8_t { TRACE, DEBUG, INFO, WARN, ERROR, FATAL };
+auto operator<<(std::ostream& /*os*/, const Level& /*level*/) -> std::ostream&;
+
+template <typename T>
+concept NotQuotable = !std::convertible_to<T, std::string> && requires(std::ostream& os, T t) {
+  { os << t } -> std::same_as<std::ostream&>;
+};
+
+template <typename T>
+concept Quotable = std::convertible_to<T, std::string>;
+
+template <typename T>
+struct Field final {
+  std::string key;
+  T val;
+};
+
+/**
+ * @brief A class that allows easy composition of logs for structured logging.
+ *        Example(see also struclog.cpp):
+ *        namespace ht=heph::telemetry;
+ *        log(ht::LogEntry{Level::INFO, "adding"} | "id"_f(12345) | ht::Field{"tag", "test"}) ;
+ *
+ *        logs
+ *        'level=info file=struclog.h:123 time=2023-12-3T8:52:02+0 message="adding" id=12345 tag="test"'
+ *
+ *        Remark: We would like to use libfmt with quoted formatting as proposed in
+ *                https://github.com/fmtlib/fmt/issues/825 once its included (instead of sstream).
+ *
+ */
+class LogEntry {
+public:
+  using Fields = std::vector<Field<std::string>>;
+  using Clock = std::chrono::system_clock;
+
+public:
+  explicit LogEntry(Level level, std::string_view message,
+                    std::source_location location = std::source_location::current());
+
+  /*
+   * @brief General loginfo consumer, should be used like LogEntry("my message") | Field{"field", 1234}
+   *        Converted to string with stringstream.
+   *
+   * @tparam T any type that is consumeable by stringstream
+   * @param key identifier for the logging data
+   * @param val value of the logging data
+   * @return LogEntry&&
+   */
+  template <NotQuotable T>
+  auto operator|(const Field<T>& field) -> LogEntry&& {
+    std::stringstream ss;
+    ss << field.val;
+    fields.emplace_back(field.key, ss.str());
+
+    return std::move(*this);
+  }
+
+  /*
+   * @brief Specialized loginfo consumer for string types so that they are quoted, should be used like
+   * LogEntry("my message") | Field{"field", "mystring"}
+   *
+   * @tparam T any type that is consumeable by stringstream
+   * @param key identifier for the logging data
+   * @param val value of the logging data
+   * @return LogEntry&&
+   */
+  template <Quotable S>
+  auto operator|(const Field<S>& field) -> LogEntry&& {
+    std::stringstream ss;
+    // Pointer decay is wanted here to catch char[n] messages
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+    ss << std::quoted(field.val);
+    fields.emplace_back(field.key, ss.str());
+
+    return std::move(*this);
+  }
+
+  Level level;
+  std::string message;
+  std::source_location location;
```

Comment:
I am thinking, given that we are going to publish the logs for all the node in a centralized space (like for metrics and influxdb) we could add the `program_name` here that we can get with something like:
```c++
auto getProgramName() -> std::string {
  /// \note This implementation only works for Linux
  char path[PATH_MAX + 1] = { '\0' }; 
  const auto path_length = readlink("/proc/self/exe", path, PATH_MAX);
  return ((path_length > 0) ? std::string(path) : (""));
}
```

Reply:
referenced in https://github.com/olympus-robotics/hephaestus/issues/202 
would do that in a spererate PR though. 

Also I am not sure about the benefit... probably with the module and the source the point where the log is happening should be clear right?

if the overhead of adding this field later on is not too big, I would postpone this to a point where we feel that we need it...